### ISSUE TITLE
QGIS crash when opening old project.

### ISSUE THREAD

**Issue by Apis49** [qgis/QGIS#46152]
### What is the bug or the crash?

Crash ID: d1f7afe8399d52daab504182ca4d3f9b8b3ed3c1

Stack Trace

QTextCodec::toUnicode :
providerMetadataFactory :
providerMetadataFactory :
providerMetadataFactory :
QgsVectorLayer::setDataProvider :
QgsVectorLayer::readXml :
QgsMapLayer::readLayerXml :
QgsProject::addLayer :
QgsProject::_getMapLayers :
QgsProject::readProjectFile :
QgsProject::unzip :
QgsProject::read :
QgisApp::addProject :
QgisApp::openProject :
QgsTemporalControllerDockWidget::temporalController :
QObject::qt_static_metacall :
QAbstractItemView::activated :
QAbstractItemView::mouseDoubleClickEvent :
QWidget::event :
QFrame::event :
QAbstractItemView::viewportEvent :
QCoreApplicationPrivate::sendThroughObjectEventFilters :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :

QGIS Info
QGIS Version: 3.22.1-Bia�owie�a
QGIS code revision: 663dcf8fb9
Compiled against Qt: 5.15.2
Running against Qt: 5.15.2
Compiled against GDAL: 3.4.0
Running against GDAL: 3.4.0

System Info
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.17134

### Steps to reproduce the issue

Open project QGIS 3.20

### Versions

3.22.1

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [x] I tried with a new QGIS profile

**Comment by gioman** [comment]
@Apis49 please try wit a new QGIS profile.

**Comment by Apis49** [comment]
I already try. It's normally opens in QGIS 3.18 but it crashes when I try to open in 3.22.1, it was the same in 3.22.0. Before I was using 3.20 without problems.

**Comment by gioman** [comment]
@Apis49 can you attach the project file?

**Comment by Apis49** [comment]
Here is the project file.

Thanks for the help.

[2021_03_23_Varstveni_sloji_PUN.zip](https://github.com/qgis/QGIS/files/7579363/2021_03_23_Varstveni_sloji_PUN.zip)

**Comment by gioman** [comment]
> Here is the project file.

@Apis49 without the data it opens fine here

![image](https://user-images.githubusercontent.com/1951107/142831511-dbfc85a4-99ab-4a7c-b1d0-4c2544b8613d.png)

**Comment by Apis49** [comment]
I try to remove the data where the loading stops. But then it stops when loading another layer. I will try on another comp.

Thank you.

**Comment by Apis49** [comment]
I managed to open the project in QGIS 3.22.1.
First I changed the name of the source map with the layers.
Then I opened the project. Of course the QGIS didn't find the layers. Then I choose AUTO-FIND layers. QGIS found layers and layers with issues were in red. So then I choose remove unavailable layers and the project opened. 

It is funny, because those layers are not corrupted. And project normally opens in older versions of QGIS. It looks that there is a problem with .XYZ, .csv and layers with filters (but not all). Maybe I am wrong. I lost enough time, so I am not digging any more.

![Zajeta slika](https://user-images.githubusercontent.com/75850744/142984095-d4c27241-6751-4f45-9083-5dac6074b50a.JPG)

![Zajeta slika2](https://user-images.githubusercontent.com/75850744/142984101-605ada9a-aab1-428c-8da6-4c299233fe60.JPG)
